### PR TITLE
store/tikv: reduce the lock contend between sending message and re-creating streaming client

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

Tiny code refactor

### What is changed and how it works?

We use a lock here
https://github.com/pingcap/tidb/compare/master...tiancaiamao:batch-client-lock?expand=1#diff-db62a1f8315185d693e2a911ce0d5b49L176
and here 
https://github.com/pingcap/tidb/compare/master...tiancaiamao:batch-client-lock?expand=1#diff-db62a1f8315185d693e2a911ce0d5b49L214

Two write-write locks to protect the send and re-create operation.
The caller of the send operation is from the `batchSendLoop`, that means the whole send loop will be blocked during the streaming client re-creating.

In this commit, I provide a `tryLockForSend` function to try the lock first, if the `batchCommandsClient` is re-creating, we can try another one, instead of blocking the `batchSendLoop`.

### Check List 

Tests 

 - No code

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @hicqu @lysu

File: store/tikv/client_batch.go [hunk 1]
```diff
@@ -365,18 +390,31 @@ func (a *batchConn) batchSendLoop(cfg config.TiKVClient) {
...
-
 		req := &tikvpb.BatchCommandsRequest{
 			Requests:   requests,
 			RequestIds: requestIDs,
 		}
 
-		batchCommandsClient.send(req, entries)
+		cli.send(req, entries)
+		tikvTransportLayerLoad = atomic.LoadUint64(cli.tikvTransportLayerLoad)
+		cli.unlockForSend()
```
Comment (lysu) [inline store/tikv/client_batch.go]: maybe better use `defer` to protect someone introduces panic in future

File: store/tikv/client_batch.go [hunk 2]
```diff
@@ -267,6 +295,7 @@ func (c *batchCommandsClient) batchRecvLoop(cfg config.TiKVClient) {
 				// TODO: Use a more smart backoff strategy.
 				time.Sleep(time.Second)
 			}
+			c.unlockForRecreate()
```
Comment (lysu) [inline store/tikv/client_batch.go]: ditto

File: store/tikv/client_batch.go [hunk 3]
```diff
@@ -224,6 +248,7 @@ func (c *batchCommandsClient) reCreateStreamingClient(err error) bool {
 			zap.String("target", c.target),
 		)
 		c.client = streamClient
```
Comment (lysu) [inline store/tikv/client_batch.go]: this write becomes a write without hold "mutex", can other thread saw this change?

Comment (tiancaiamao) [reply store/tikv/client_batch.go]: `reCreateStreamingClient` acquire the write lock, so other send goroutine skips this `batchCommandsClient` and try the next one, in `tryLockForSend` @lysu

File: store/tikv/client_batch.go (hunk 1 above)
Comment (tiancaiamao) [reply store/tikv/client_batch.go]: Done

File: store/tikv/client_batch.go (hunk 2 above)
Comment (tiancaiamao) [reply store/tikv/client_batch.go]: Done

File: store/tikv/client_batch.go (hunk 3 above)
Comment (lysu) [reply store/tikv/client_batch.go]: `lockForRecreate` set a flag to exclusive others to step into this, but it call mutex.Unlock after set flag, so this function can exclusive other access, but I still confuse its visibility :confused:

**Comment by tiancaiamao** [comment]
PTAL @lysu @hicqu

File: store/tikv/client_batch.go (hunk 3 above)
Comment (tiancaiamao) [reply store/tikv/client_batch.go]: After `lockForRecreate`, the `reCreate` flag is set.  `tryLockForSend` will detect the flag.
Maybe you can implement a `RWMutex` struct with a `tryRlock` API, and then you'll figure out it.

File: store/tikv/client_batch.go [hunk 4]
```diff
@@ -341,7 +374,6 @@ func (a *batchConn) batchSendLoop(cfg config.TiKVClient) {
 		a.fetchAllPendingRequests(int(cfg.MaxBatchSize), &entries, &requests)
 
 		if len(entries) < int(cfg.MaxBatchSize) && cfg.MaxBatchWaitTime > 0 {
-			tikvTransportLayerLoad := atomic.LoadUint64(batchCommandsClient.tikvTransportLayerLoad)
```
Comment (hicqu) [inline store/tikv/client_batch.go]: You can get `tikvTransportLayerLoad` from `batchConn` instead of `batchClmmandsClient`.

**Comment by hicqu** [comment]
rest LGTM.

File: store/tikv/client_batch.go (hunk 4 above)
Comment (tiancaiamao) [reply store/tikv/client_batch.go]: Done

File: store/tikv/client_batch.go [hunk 5]
```diff
@@ -426,7 +425,7 @@ func (a *batchConn) getClientAndSend(entries []*batchCommandsEntry, requests []*
 	}
 
 	cli.send(req, entries)
-	return atomic.LoadUint64(cli.tikvTransportLayerLoad)
+	return atomic.LoadUint64(&a.tikvTransportLayerLoad)
```
Comment (hicqu) [inline store/tikv/client_batch.go]: I guess it's not necessary any more?

Comment (tiancaiamao) [reply store/tikv/client_batch.go]: Done

**Pull request merged**